docs(theming): Adds details around new theme and glass mode. (#4983)
patternfly/patternfly-org: packages/documentation-site/patternfly-docs/content/foundations-and-styles/styles/theming/glass-mode-handbook.md

---
id: Theming
section: foundations-and-styles
source: glass-mode-handbook
---

# Glass mode developer handbook 

## What is glass mode?

Glass mode is a contrast mode option that's enabled in our Project Felt theme by default and can be manually enabled in our Default theme. Glass adds transparency, blurring, and depth to the UI to create a more dynamic visual effect where brand-approved background images and layered UI elements subtly show through. 

When turned on, glass applies to the following components, including their uses within our extensions:
- Card
- Drawer
- Login page 
- Masthead

> - Navigation (docked)

Reviewer: The left navigation gets a glass panel as well for contrast.

<img width="761" height="502" alt="Image" src="https://github.com/user-attachments/assets/83969ae0-9cb0-405e-9aee-d820e7744293" />

Author: I just removed "(docked)" to just state Navigation. Does that seem okay?

## Enabling glass mode

Glass mode is designed to work across light and dark color schemes and is enabled by default in our Project Felt theme. To manually enable glass in our Default theme, add the class `.pf-v6-theme-glass` to your application’s `<html>` tag. When implementing glass with our Default theme, it's important to ensure that glass does not harm the overall accessibility or usability of your product.

### Technical constraints

There are a few technical constraints to abide by when using glass in your product.

- **No glass-on-glass layering:** Never layer glass-enabled containers. Doing so can cause significant performance and accessibility problems. For example, child objects inside a blurred parent container can become illegibly blurred themselves. To mitigate this, we have intentionally adjusted the opacity of our background color design tokens to simulate depth without introducing extra blur.
- **High contrast precedence:** If high-contrast mode is enabled, all glass effects must be automatically disabled to prioritize functional accessibility.

> - **User controls and preferences:** Because glass mode introduces legibility risks, any product utilizing glass must also let users swap between default contrast and high contrast modes via a theme switcher or preferences menu. Products should also respect the OS-level `prefers-reduced-transparency` media query, disabling glass or replacing it with a solid high-opacity background to accommodate users appropriately. 

Reviewer: Do we want to say something like "Because glass mode may introduce legibility issues for some users..."
Trying to make it sound less like an expected outcome but that;s the best I got!

Author: Fair point - went with this? 

> Because some users might experience legibility issues in glass mode, any product utilizing glass must also let users swap to default contrast or high contrast via a theme switcher or preferences menu.

## Glass design tokens 

The following tokens are used together to create the glass effect.

| Token | Value (in Project Felt theme) | Usage |
| --- | --- | --- |

> | `-pf-t--global--background--color--glass--primary--default` | Light mode value: #FFFFFF (60% opacity) <br />  Dark mode value: #292929 at 50% opacity | Base fill for glass containers. |

Reviewer: Not certain but in the code diff panel it looks like the double hyphens autocorrected to an en dash?

Author: ah that was a me-typo

---
id: Theming
section: foundations-and-styles
source: glass-mode-handbook
---

# Glass mode developer handbook 

## What is glass mode?

> 
> Glass mode is a contrast mode option that's enabled in our Project Felt theme by default and can be manually enabled in our Default theme. Glass adds transparency, blurring, and depth to the UI to create a more dynamic visual effect where brand-approved background images and layered UI elements subtly show through. 
> 
> When turned on, glass applies to the following components, including their uses within our extensions:
> - Card
> - Drawer
> - Login page 
> - Masthead
> - Navigation
> - Page
> - Panel

Author: ```suggestion
When turned on, glass applies to the following components, including their uses within our extensions:
- Login page
- Masthead
- Navigation
- Page

Additionally, glass can be manually enabled on the following components, via the `isGlass` prop: 
- Card
- Drawer
- Panel
```

- Drawer

> - Panel

Reviewer 2: hero can also get glass!

### Opacity

The default opacity values used in our components have been tested for accessibility, legibility, and visual appeal. Glass surfaces (page, masthead, navigation, and similar components) use 50% opacity in both light and dark mode, with a background blur effect applied.

If a product team chooses to override these token values, they are responsible for conducting their own accessibility evaluation to ensure WCAG compliance.

## Enabling glass mode

Glass mode is designed to work across light and dark color schemes in both our Default and Project Felt themes. To enable glass, add the class `.pf-v6-theme-glass` to your application’s `<html>` tag. When implementing glass, it’s important to ensure that it does not harm the overall accessibility or usability of your product.

### Technical constraints

There are a few technical constraints to abide by when using glass in your product.

> - **No glass-on-glass layering:** Never layer glass-enabled containers. Doing so can cause significant performance and accessibility problems. For example, child objects inside a blurred parent container can become illegibly blurred themselves. To mitigate this, we have intentionally adjusted the opacity of our background color design tokens to simulate depth without introducing extra blur.

Reviewer 2: might be worth noting that we have introduced opaque styles for any items that are floating directly on top of glass elements (i.e. sticky sections, menus, popovers, overlay drawers, etc.)

- **High contrast precedence:** If high-contrast mode is enabled, all glass effects must be automatically disabled to prioritize functional accessibility.
- **User controls and preferences:** Because some users might experience legibility issues in glass mode, any product utilizing glass must also let users swap to default contrast or high contrast via a theme switcher or preferences menu. Products should also respect the OS-level `prefers-reduced-transparency` media query, disabling glass or replacing it with a solid high-opacity background to accommodate users appropriately. 

## Glass design tokens 

The following tokens are used together to create the glass effect.

| Token | Value (in Project Felt theme) | Usage |
| --- | --- | --- |

> | `--pf-t--global--background--color--glass--primary--default` | Light mode value: #FFFFFF (60% opacity) <br />  Dark mode value: #292929 at 50% opacity | Base fill for glass containers. |

Reviewer 2: 50% in both themes.

> | `--pf-t--global--background--filter--glass--default` | Light mode value: #C7C7C7 (25% opacity) <br />  Dark mode value: #C7C7C7 (15% opacity) | Overlay tint used to enhance blur visibility. |

Reviewer 2: we dropped this to a lower 16% blur in both themes. also this is not an overlay tint, its the actual blur value!

> | `--pf-t--global--border--color--glass--default` | Light mode value: #FFFFFF (60% opacity) <br />  Dark mode value: #292929 (50% opacity) | Boundary highlight for glass surfaces. |

Reviewer 2: this references our border--color--alt token in glass mode (which maps to primary bkg color)

When turned on, glass applies to the following components, including their uses within our extensions:
- Login page
- Masthead
- Navigation
- Page

Additionally, glass can be manually enabled on the following components, via the `isGlass` prop: 
- Card
- Drawer
- Panel

### Background images

The glass effect is most visible when placed over a background image. To retain readability and ensure proper contrast ratios are met, images shouldn't contain high levels of detail or extreme contrast. 

Background images should be selected for each Red Hat product in collaboration with the Brand team.

Text must never be placed directly on a background image&mdash;it should be placed within a container that has a background color or glass effect. Titles or headings with stronger font weights can be placed directly on background images only if they pass specific brand and contrast requirements.

### Opacity

The default opacity values used in our components have been tested for accessibility, legibility, and visual appeal.

> - Project Felt light typically utilizes a 40% opacity to maintain AAA contrast for standard text.

Reviewer: Glass surfaces like page, masthead, navigation, etc use 50% opacity in both light and dark mode. They also have the background blur effect applied.

Author: I updated this section to be this: "Glass surfaces (page, masthead, navigation, and similar components) use 50% opacity in both light and dark mode, with a background blur effect applied."

I feel like this statement conflicts with what I put in this token table - is this not accurate? I tried pulling these values from the unified theme file, but might have not done so correctly 

| Token | Value (in Project Felt theme) | Usage |
| --- | --- | --- |
| `--pf-t--global--background--color--glass--primary--default` | Light mode value: #FFFFFF (60% opacity) <br />  Dark mode value: #292929 at 50% opacity | Base fill for glass containers. |
| `--pf-t--global--background--filter--glass--default` | Light mode value: #C7C7C7 (25% opacity) <br />  Dark mode value: #C7C7C7 (15% opacity) | Overlay tint used to enhance blur visibility. |
| `--pf-t--global--border--color--glass--default` | Light mode value: #FFFFFF (60% opacity) <br />  Dark mode value: #292929 (50% opacity) | Boundary highlight for glass surfaces. |
| `--pf-t--global--border--radius--glass--default` |  #D6D6D6 with 16px radius | Default rounded border for glass element. |
| `--pf-t--global--background--color--sticky--default` | Light mode value: #FFFFFF (100% opacity) <br />  Dark mode value: #292929 (100% opacity) | Solid background for sticky elements, like headers. |
| `--pf-t--global--box-shadow--medium` | #292929 (15% opacity) | Shadow that signifies elevation. |
